README.md
# mdns-lookup
Python mDNS lookup

[![Github Actions Tests](https://github.com/AndreMiras/mdns-lookup/workflows/Tests/badge.svg)](https://github.com/AndreMiras/mdns-lookup/actions?query=workflow%3ATests)
[![PyPI version](https://badge.fury.io/py/mdns-lookup.svg)](https://badge.fury.io/py/mdns-lookup)

This is based on the Stackoverflow answer http://stackoverflow.com/a/35853322/185510.

## Usage
Lookup on multiple hostnames:
```sh
mdnslookup raspberrypi.local hprinter.local
```
Output:
```text
raspberrypi.local 192.168.1.24
hprinter.local 192.168.1.42
```

## Install
```sh
pip install mdns-lookup
```

## Tests
```sh
tox
```


CHANGELOG.md
# Change Log

## [v20200613]
- Automate PyPI releases from CI
- Introduce isort and black code formatter
- Add `mdnslookup` script entrypoint

## [v20200612]
- Setup continuous integration
- Add Python 3 support, refs #1

## [v20160930]
- Added unit tests

## [v20160809]
- Initial release


test_mdnslookup.py
"""
Verifies core features run as expected.
"""
from __future__ import print_function
import sys
import unittest
from contextlib import contextmanager
from mdnslookup import lookup, format_results
try:
    from StringIO import StringIO
except ImportError:
    from io import StringIO


@contextmanager
def captured_output():
    new_out, new_err = StringIO(), StringIO()
    old_out, old_err = sys.stdout, sys.stderr
    try:
        sys.stdout, sys.stderr = new_out, new_err
        yield sys.stdout, sys.stderr
    finally:
        sys.stdout, sys.stderr = old_out, old_err


class TestCore(unittest.TestCase):

    def test_lookup(self):
        """
        Verifies the lookup() function returns a dict as expected.
        """
        hosts_list = [
            # testing one host
            ["raspberrypi.local"],
            # testing two hosts
            ["raspberrypi.local", "whatever.local"],
        ]
        for hosts in hosts_list:
            hosts_ip = lookup(hosts)
            self.assertEqual(len(hosts_ip), len(hosts))
            self.assertEqual(set(hosts_ip.keys()), set(hosts))

    def test_format_results(self):
        """
        Verifies the format_results() correctly prints the results.
        """
        hosts_ip = {
            'raspberrypi.local': '192.168.1.116',
            'whatever.local': None
        }
        with captured_output() as (out, err):
            format_results(hosts_ip)
            output = out.getvalue()
            lines = output.split("\n")
            self.assertTrue(
                'raspberrypi.local' in lines[0])
            self.assertTrue(
                'whatever.local' in lines[1])

if __name__ == '__main__':
    unittest.main()


tests/test_mdnslookup.py
"""
Verifies core features run as expected.
"""
from __future__ import print_function

import sys
import unittest
from contextlib import contextmanager

from mdnslookup import format_results, lookup

try:
    from StringIO import StringIO
except ImportError:
    from io import StringIO


@contextmanager
def captured_output():
    new_out, new_err = StringIO(), StringIO()
    old_out, old_err = sys.stdout, sys.stderr
    try:
        sys.stdout, sys.stderr = new_out, new_err
        yield sys.stdout, sys.stderr
    finally:
        sys.stdout, sys.stderr = old_out, old_err


class TestCore(unittest.TestCase):
    def test_lookup(self):
        """
        Verifies the lookup() function returns a dict as expected.
        """
        hosts_list = [
            # testing one host
            ["raspberrypi.local"],
            # testing two hosts
            ["raspberrypi.local", "whatever.local"],
        ]
        for hosts in hosts_list:
            hosts_ip = lookup(hosts)
            self.assertEqual(len(hosts_ip), len(hosts))
            self.assertEqual(set(hosts_ip.keys()), set(hosts))

    def test_format_results(self):
        """
        Verifies the format_results() correctly prints the results.
        """
        hosts_ip = {"raspberrypi.local": "192.168.1.116", "whatever.local": None}
        with captured_output() as (out, err):
            format_results(hosts_ip)
            output = out.getvalue()
            for host in hosts_ip.keys():
                self.assertTrue(host in output)


if __name__ == "__main__":
    unittest.main()


lookup.py
#!/usr/bin/env python
"""
Python lookup module and program.
Inspired from the stackoverflow answer:
http://stackoverflow.com/a/35853322/185510
"""
from __future__ import print_function
import argparse
import socket
import struct
from dpkt.dns import DNS, DNS_A

VERBOSE_GLOBAL = False


def lookup(hosts):
    UDP_IP = "0.0.0.0"
    UDP_PORT = 5353
    MCAST_GRP = '224.0.0.251'
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.bind((UDP_IP, UDP_PORT))
    # join the multicast group
    mreq = struct.pack("4sl", socket.inet_aton(MCAST_GRP), socket.INADDR_ANY)
    sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, mreq)

    for host in hosts:
        # the string in the following statement is an empty query packet
        dns = DNS(
                '\x00\x00\x01\x00\x00\x01\x00\x00' +
                '\x00\x00\x00\x00\x00\x00\x01\x00\x01')
        dns.qd[0].name = host
        sock.sendto(dns.pack(), (MCAST_GRP, UDP_PORT))
    sock.settimeout(5)
    while True:
        try:
            m = sock.recvfrom(1024)
            if VERBOSE_GLOBAL:
                print('%r %s' % (m[0], m[1]))
            dns = DNS(m[0])
            if len(dns.qd) > 0:
                print("%r %s" % (dns, dns.qd[0].name)
            if len(dns.an) > 0 and dns.an[0].type == DNS_A:
                print("%r %s %s" % (
                        dns,
                        dns.an[0].name,
                        socket.inet_ntoa(dns.an[0].rdata)))
        except socket.timeout:
            break


def main():
    """
    Parses sys args and calls the lookup function.
    """
    parser = argparse.ArgumentParser(description='Performs mDNS lookups.')
    parser.add_argument(
            '-v', '--verbose', help='Adds output verbosity')
    parser.add_argument(
            'hosts', nargs='+', help='Hosts to resolve')
    args = parser.parse_args()
    global VERBOSE_GLOBAL
    VERBOSE_GLOBAL = args.verbose
    hosts = args.hosts
    lookup(hosts)

if __name__ == "__main__":
    main()


mdnslookup/mdnslookup.py
#!/usr/bin/env python
"""
Python lookup module and program.
Inspired from the stackoverflow answer:
http://stackoverflow.com/a/35853322/185510
"""
from __future__ import print_function

import argparse
import socket
import struct

from dpkt.dns import DNS, DNS_A

VERBOSE_GLOBAL = False


def lookup(hosts):
    """
    Returns a host/ip dictionary for the given host list.
    """
    hosts_ip = dict.fromkeys(hosts)
    UDP_IP = "0.0.0.0"
    UDP_PORT = 5353
    MCAST_GRP = "224.0.0.251"
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.bind((UDP_IP, UDP_PORT))
    # join the multicast group
    mreq = struct.pack("4sl", socket.inet_aton(MCAST_GRP), socket.INADDR_ANY)
    sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, mreq)

    for host in hosts:
        # the string in the following statement is an empty query packet
        dns = DNS(
            b"\x00\x00\x01\x00\x00\x01\x00\x00"
            + b"\x00\x00\x00\x00\x00\x00\x01\x00\x01"
        )
        dns.qd[0].name = host
        sock.sendto(dns.pack(), (MCAST_GRP, UDP_PORT))
    # receives until the end of the timeout
    sock.settimeout(5)
    while True:
        try:
            m = sock.recvfrom(1024)
            if VERBOSE_GLOBAL:
                print("%r %s" % (m[0], m[1]))
            dns = DNS(m[0])
            if len(dns.qd) > 0:
                if VERBOSE_GLOBAL:
                    print("%r %s" % (dns, dns.qd[0].name))
            if len(dns.an) > 0 and dns.an[0].type == DNS_A:
                host = dns.an[0].name
                ip = socket.inet_ntoa(dns.an[0].rdata)
                if VERBOSE_GLOBAL:
                    print("%r %s %s" % (dns, host, ip))
                hosts_ip[host] = ip
        except socket.timeout:
            break
    return hosts_ip


def format_results(host_ip):
    """
    Format the result to the output in the form of:
    host ip
    """
    for host, ip in host_ip.items():
        print("%s %s" % (host, ip))


def main():
    """
    Parses sys args and calls the lookup function.
    """
    parser = argparse.ArgumentParser(description="Performs mDNS lookups.")
    parser.add_argument("-v", "--verbose", default=False, help="Adds output verbosity")
    parser.add_argument("hosts", nargs="+", help="Hosts to resolve")
    args = parser.parse_args()
    global VERBOSE_GLOBAL
    VERBOSE_GLOBAL = args.verbose
    hosts = args.hosts
    hosts_ip = lookup(hosts)
    format_results(hosts_ip)


if __name__ == "__main__":
    main()
